### PR TITLE
Profiling and misc

### PR DESCRIPTION
**Pull request by jlamypoirier**
* Add GPU profiling
* Ultra-fast model creation so we can benchmark/profile quickly. The model is created in ~300 ms (create and init weights directly on GPU) instead of 20-30 s loading + several minutes on the first time. On the downside, we no longer benchmark the loading time and can't use custom weights but that could be added easily (and it would be much faster than before).
* Rewrite the benchmarking code and pipeline initialization to support the new features.
* Use the logging module instead of plain prints and add timestamps so we can more easily track the progress
* Add some safety checks on arguments
* Improve performance metrics and add several more
* Simplify device placement (with arg instead of separate class) and data types
* Move ds/hf common code to the `Pipeline` class
* Lots of refactoring and cleanup

### CODE REVIEW

File: src/pipelines/pipeline.py [hunk 1]
```diff
@@ -1,27 +1,112 @@
...
+        self.tokenizer = AutoTokenizer.from_pretrained("gpt2")
+        self.tokenizer.add_special_tokens({"pad_token": "[PAD]"})
+
+        self.device = args.device
+
+        model_class, config = self.get_config(args)
+        is_int8 = args.dtype == torch.int8
+        if is_int8:
+            check_unused(args, {"device": torch.device("cuda")}, enforce=True)
+        torch_dtype = torch.float16 if is_int8 else args.dtype
```
Comment (mayank31398) [inline src/pipelines/pipeline.py:45]: shouldn't this be torch.int8?

Comment (jlamypoirier) [reply src/pipelines/pipeline.py:45]: We can't create the model directly in int8, we have to go through a float.

File: Makefile [hunk 2]
```diff
@@ -16,59 +16,43 @@ install-mqa-transformers:
 
 # BLOOM AliBi
 hf-1b-bloom-fp32:
-	python src/main.py --hidden_size 2048 --n_head 16 --n_layer 24 --pipeline_class HF_GPU_Pipeline --model_class BLOOM --dtype float32 --batch_size ${batch_size}
+	python3 src/main.py --hidden_size 2048 --n_head 16 --n_layer 24 --pipeline_class HF_Pipeline --model_class BLOOM --dtype float32 --batch_size ${batch_size}
```
Comment (jlamypoirier) [inline Makefile:15]: @mayank31398 I updated the targets and removed the obsolete ones, but we should move them away from the makefile in favour of bundling them with the experiments.

Comment (mayank31398) [reply Makefile:15]: Agreed, will ditch it over time

File: src/pipelines/pipeline.py (hunk 1 above)
Comment (mayank31398) [reply src/pipelines/pipeline.py:45]: makes sense. yeah, my bad. Totally forgot about that

File: src/pipelines/pipeline.py [hunk 3]
```diff
@@ -1,27 +1,113 @@
...
+        t0 = time.perf_counter()
         input_tokens = self.tokenizer(text, return_tensors="pt", padding=True)
 
         for t in input_tokens:
             if torch.is_tensor(input_tokens[t]):
-                input_tokens[t] = input_tokens[t].to(self.input_device)
+                input_tokens[t] = input_tokens[t].to(self.device)
 
+        t1 = time.perf_counter()
         with torch.no_grad():
```
Comment (mayank31398) [inline src/pipelines/pipeline.py:107]: Is this needed?
The generate method already runs in torch.no_grad context by default

**Review by mayank31398 — APPROVED** [review]
Added a few changes. Most of it looks good to me :)

File: src/pipelines/pipeline.py (hunk 3 above)
Comment (jlamypoirier) [reply src/pipelines/pipeline.py:107]: It's probably useless but a bit of extra safety can't hurt. Same with the `model.eval()` in `__init__`.

Comment (njhill) [reply src/pipelines/pipeline.py:107]: Any reason not to use `inference_mode()`?

Comment (jlamypoirier) [reply src/pipelines/pipeline.py:107]: I didn't know about this one, thanks for pointing it out. I will leave it for future work though.

Comment (jlamypoirier) [reply src/pipelines/pipeline.py:107]: #12

Comment (njhill) [reply src/pipelines/pipeline.py:107]: Sure, I'm not sure of the context here but it should be a harmless replacement ... it includes everything no_grad does.

File: Makefile [hunk 4]
```diff
@@ -6,69 +6,49 @@ style:
...
-	cd ..; \
-	rm -rf transformers;
+install:
+	git submodule update --init
+	pip install -r requirements.txt
 
 # BLOOM AliBi
 hf-1b-bloom-fp32:
-	python src/main.py --hidden_size 2048 --n_head 16 --n_layer 24 --pipeline_class HF_GPU_Pipeline --model_class BLOOM --dtype float32 --batch_size ${batch_size}
+	python3 src/main.py --hidden_size 2048 --n_head 16 --n_layer 24 --pipeline_class HF_Pipeline --model_class BLOOM --dtype float32 --batch_size ${batch_size}
```
Comment (jlamypoirier) [inline Makefile:15]: @mayank31398 I don't see why we should have it as a module. It's a script and not installed as a package.

File: src/pipelines/pipeline.py [hunk 5]
```diff
@@ -30,48 +115,36 @@ def __call__(self, text: List[str], **generate_kwargs) -> Tuple[List[str], List[
...
-        )
-        model_class = GPT2LMHeadModel
-
-    if not os.path.exists("tmp"):
-        model_class._from_config(config).save_pretrained("tmp")
-
-    return config, tokenizer, model_class
+        return sum(p.numel() for p in self.model.parameters())
+
+    def aggregate_and_format_metrics(self, metrics: List[Dict[str, Any]]):
```
Comment (jlamypoirier) [inline src/pipelines/pipeline.py:133]: @mayank31398 It's better to leave this encapsulated with the class that generates these metrics.

**Pull request merged**